File: backend/src/main/java/com/lingflow/entity/ProcessCategory.java
package com.lingflow.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

/**
 * 流程分类实体类
 * 支持多租户、树形结构、路径记录
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ProcessCategory {
    /**
     * 主键ID (UUID)
     */
    private String id;

    /**
     * 自动生成UUID并在构造函数中设置
     */
    public ProcessCategory(String id) {
        this.id = id != null ? id : UUID.randomUUID().toString();
    }

    /**
     * 分类名称
     */
    private String name;

    /**
     * 分类编码（用于API调用、系统引用等）
     * 同一租户、应用、上下文下唯一
     */
    private String code;

    /**
     * 父分类ID (NULL表示根节点)
     */
    private String parentId;

    /**
     * 分类路径（格式：/id1/id2/id3）
     * 用于快速查询某个节点下的所有后代
     */
    private String path;

    /**
     * 层级深度（根节点为0）
     */
    private Integer level;

    /**
     * 分类描述
     */
    private String description;

    /**
     * 排序序号（同级分类的排序）
     */
    private Integer sortOrder;

    /**
     * 图标（Element Plus 图标名或 URL）
     */
    private String icon;

    // ============ 多租户字段 ============

    /**
     * 应用ID（可选，用于更细粒度的隔离）
     */
    private String appId;

    /**
     * 上下文ID（可选，用于特定业务场景）
     */
    private String contextId;

    /**
     * 租户ID（必需，用于多租户隔离）
     */
    private String tenantId;

    // ============ 审计字段 ============

    /**
     * 创建时间
     */
    private LocalDateTime createdTime;

    /**
     * 更新时间
     */
    private LocalDateTime updatedTime;

    /**
     * 创建人
     */
    private String createdBy;

    /**
     * 更新人
     */
    private String updatedBy;

    /**
     * 软删除标记
     */
    private Boolean isDeleted;

    /**
     * 乐观锁版本号
     */
    private Integer version;

    // ============ 非持久化字段 ============

    /**
     * 子分类列表（用于树形结构）
     * 不持久化到数据库，仅用于查询结果
     */
    private List<ProcessCategory> children;

    /**
     * 该分类下的流程数量
     * 不持久化到数据库，仅用于查询结果
     */
    private Integer processCount;

    /**
     * 是否有子分类
     * 不持久化到数据库，仅用于查询结果
     */
    private Boolean hasChildren;
}

File: frontend/src/api/processCategory.ts
import request from './index'

// 流程分类接口
export interface ProcessCategory {
  id?: string
  name: string
  code: string
  parentId?: string
  path?: string
  level?: number
  description?: string
  sortOrder?: number
  icon?: string
  appId?: string
  contextId?: string
  tenantId: string
}

// 流程分类树接口
export interface ProcessCategoryTree extends ProcessCategory {
  processCount?: number
  hasChildren?: boolean
  children?: ProcessCategoryTree[]
}

// 流程搜索结果接口
export interface ProcessSearchResult {
  id: string
  key: string
  name: string
  version: number
  categoryId: string
  categoryName: string
  categoryPath: string         // 路径ID：root/id1/id2
  categoryPathNames: string    // 路径名称：根节点/子节点1/子节点2
  pathIds: string[]           // 用于前端定位
}

// 排序数据接口
export interface SortOrderDTO {
  id: string
  sortOrder: number
}

/**
 * 获取分类树
 */
export function getCategoryTree(params: {
  tenantId: string
  appId?: string
  contextId?: string
}) {
  return request({
    url: '/api/process-category/tree',
    method: 'get',
    params
  })
}

/**
 * 创建分类
 */
export function createCategory(data: ProcessCategory) {
  return request({
    url: '/api/process-category',
    method: 'post',
    data
  })
}

/**
 * 更新分类
 */
export function updateCategory(id: string, data: ProcessCategory) {
  return request({
    url: `/api/process-category/${id}`,
    method: 'put',
    data
  })
}

/**
 * 删除分类
 */
export function deleteCategory(id: string) {
  return request({
    url: `/api/process-category/${id}`,
    method: 'delete'
  })
}

/**
 * 移动分类
 */
export function moveCategory(id: string, newParentId: string) {
  return request({
    url: `/api/process-category/${id}/move`,
    method: 'put',
    params: { newParentId }
  })
}

/**
 * 更新排序
 */
export function updateSortOrder(id: string, sortOrder: number) {
  return request({
    url: `/api/process-category/${id}/sort`,
    method: 'put',
    params: { sortOrder }
  })
}

/**
 * 批量更新排序（拖拽排序）
 */
export function batchUpdateSortOrder(sortOrderList: SortOrderDTO[]) {
  return request({
    url: '/api/process-category/sort/batch',
    method: 'put',
    data: sortOrderList
  })
}

/**
 * 搜索分类
 */
export function searchCategories(params: {
  keyword: string
  tenantId: string
  appId?: string
  contextId?: string
}) {
  return request({
    url: '/api/process-category/search',
    method: 'get',
    params
  })
}

/**
 * 设置流程分类
 */
export function setProcessCategory(
  processDefinitionId: string,
  categoryId: string,
  params: {
    tenantId: string
    appId?: string
    contextId?: string
  }
) {
  return request({
    url: `/api/process-category/process/${processDefinitionId}`,
    method: 'put',
    params: { categoryId, ...params }
  })
}

/**
 * 搜索流程模板（返回分类路径）
 */
export function searchProcesses(params: {
  keyword: string
  categoryId?: string
  tenantId: string
  appId?: string
  contextId?: string
}) {
  return request({
    url: '/api/process-category/processes/search',
    method: 'get',
    params
  })
}
